### PR TITLE
Ensuring operator updates ScaledObject on all HTTPScaledObject changes

### LINKED ISSUE #20: Make the operator "maintain" all the app resources it creates
Currently if you create an `HTTPScaledObject`, the operator will create a bunch of Kubernetes resources to encompass the "app" - things like `Service`s, `Deployment`s, `ScaledObject`s...

If you delete one of those resources, though, the app won't work properly and the operator isn't aware of that. When you go to delete or edit (https://github.com/kedacore/http-add-on/issues/7) the app, the resource just won't be there and the operator will just fail silently (on deletes, it just logs the missing resource and moves on).

It would be most helpful for the operator to somehow ensure the resources don't get deleted. I think this could be done with a finalizer on each of them.

### Use-Case

I'd like to more tightly "tie" the app resources to the `HTTPScaledObject` that caused them to be created, so that:

- Resources are more difficult to delete without also deleting the `HTTPScaledObject`
- If a resource is deleted, the operator will know about it and can (theoretically) re-create it

### Specification

- [ ] Demand #1

### LINKED ISSUE #323: httpscaledobjects does not trigger a change in the corresponding scaledobjects.
The httpscaledobjects resource does not trigger a change in the corresponding scaledobjects resource.(I'm not sure if this is a bug)

## Expected Behavior
When I update an HSO resource, I want the corresponding SO resource to be updated as well.

## Actual Behavior
When I update the HSO resource, the corresponding SO resource is not updated.

## Steps to Reproduce the Problem

  1. Create a HSO

     ```yaml
     kind: HTTPScaledObject
     apiVersion: http.keda.sh/v1alpha1
     metadata:
         name: xkcd
         namespace: keda
     spec:
         host: myhost.com
         scaleTargetRef:
             deployment: xkcd
             service: xkcd
             port: 8080
         replicas:
             min: 0
             max: 10
     ```

     The HSO and SO resource have been created.

     ```shell
     # kubectl get httpscaledobjects.http.keda.sh -n keda
     NAME   SCALETARGETDEPLOYMENTNAME   SCALETARGETSERVICENAME                               SCALETARGETPORT   MINREPLICAS   MAXREPLICAS   AGE    ACTIVE
     xkcd                               {"deployment":"xkcd","port":8080,"service":"xkcd"}                     0             10            173m
     
     # kubectl get scaledobjects.keda.sh -n keda
     NAME                            SCALETARGETKIND      SCALETARGETNAME                 MIN   MAX   TRIGGERS        AUTHENTICATION   READY   ACTIVE   FALLBACK   AGE
     keda-add-ons-http-interceptor   apps/v1.Deployment   keda-add-ons-http-interceptor   1     50    external                         True    True     False      3h19m
     xkcd-app                        apps/v1.Deployment   xkcd                            0     10    external-push                    True    False    False      173m
     ```

2. Update the xkcd HSO with the following configuration.

   ```yaml
   kind: HTTPScaledObject
   apiVersion: http.keda.sh/v1alpha1
   metadata:
       name: xkcd
       namespace: keda
   spec:
       host: myhost.com
       scaleTargetRef:
           deployment: xkcd
           service: xkcd
           port: 8080
       replicas:
           min: 5 # change spec.replicas.min to 5
           max: 10
   ```

   The HSO resource has been updated, but the SO resource has not.

   ```shell
   # kubectl get httpscaledobjects.http.keda.sh -n keda
   NAME   SCALETARGETDEPLOYMENTNAME   SCALETARGETSERVICENAME                               SCALETARGETPORT   MINREPLICAS   MAXREPLICAS   AGE    ACTIVE
   xkcd                               {"deployment":"xkcd","port":8080,"service":"xkcd"}                     5             10            177m
   
   # kubectl get scaledobjects.keda.sh -n keda
   NAME                            SCALETARGETKIND      SCALETARGETNAME                 MIN   MAX   TRIGGERS        AUTHENTICATION   READY   ACTIVE   FALLBACK   AGE
   keda-add-ons-http-interceptor   apps/v1.Deployment   keda-add-ons-http-interceptor   1     50    external                         True    True     False      3h23m
   xkcd-app                        apps/v1.Deployment   xkcd                            0     10    external-push                    True    False    False      177m
   ```

## Specifications

- **KEDA Version:** *2.4.0*
- **KEDA HTTP Add-on Version:** *0.2.0.RC1*
- **Platform & Version:** *Ubuntu 18.04 LTS x86_64*
- **Kubernetes Version:** *v1.21.5*

I noticed that during the HSO reconciliation process, the corresponding SO resource is created via [createScaledObjects()](https://github.com/kedacore/http-add-on/blob/dda17c5eaec7179ea5849d9e3b1d626f3dbc7849/operator/controllers/scaled_object.go#L16). But why not to update the SO when it already exists?

### PR DESCRIPTION
**Pull request by arschles**
In certain situations, a valid update to an `HTTPScaledObject` won't trigger any in-kind update to the corresponding `ScaledObject`

### Checklist

- [x] Test in live Kubernetes
- [x] Commits are signed with Developer Certificate of Origin (DCO)
- [x] Any necessary documentation is added, such as:
  - [`README.md`](/README.md)
  - [The `docs/` directory](./docs)
  - [The docs repo](https://github.com/kedacore/keda-docs)

Fixes #323
Fixes #20

### CODE REVIEW

**Comment by arschles** [comment]
@tpiperatgod could you take a look at this when you have a chance? It's almost ready but I would appreciate your review.

**Review by bketelsen — APPROVED** [review]
LGTM

**Review by tpiperatgod — APPROVED** [review]

**Comment by tpiperatgod** [comment]
> @tpiperatgod could you take a look at this when you have a chance? It's almost ready but I would appreciate your review.

Sorry, I'm just now seeing this, I think it will solve my issue, thanks @arschles !

File: pkg/k8s/scaledobject.go [hunk 1]
```diff
@@ -7,31 +7,61 @@ import (
...
+}
+
 // DeleteScaledObject deletes a scaled object with the given name
-func DeleteScaledObject(ctx context.Context, name string, namespace string, cl client.Client) error {
-	scaledObj := &unstructured.Unstructured{}
+func DeleteScaledObject(
+	ctx context.Context,
+	ns,
+	name string,
+	cl client.Client,
```
[the remark this answers is not in the document]
Comment (arschles) [reply pkg/k8s/scaledobject.go]: good idea. thanks, fixed in https://github.com/kedacore/http-add-on/pull/324/commits/243be6ce3cdc5e398357c1a114a3b607c662d17d

**Comment by arschles** [comment]
I screwed up a rebase on this PR. Closing in favor of the smaller #402.

**Pull request closed without merging**